overlord/devicestate/devicestate_test.go

```diff
@@ -2408,3 +2418,278 @@ func (s *deviceMgrSuite) TestEnsureExpiredUsersRemovedNotUnseeded(c *C) {
+}
+
+func (s *deviceMgrSuite) TestHandleAutoImportAssertionNoSeedCache(c *C) {
+	a := devicestate.MockProcessAutoImportAssertion(func(*state.State, seed.Seed, asserts.RODatabase, func(batch *asserts.Batch) error) error {
+		panic("trying to process auto-import-assertion without cached system seed")
+	})
+	defer a()
+
+	s.mockSystemMode(c, "run")
+
+	devicestate.EnsureAutoImportAssertions(s.mgr)
+
+	// ensure state has not been changed
+	s.state.Lock()
+	defer s.state.Unlock()
+	var asState string
+	err := s.state.Get("auto-as-imported", &asState)
+	c.Assert(err, testutil.ErrorIs, state.ErrNoState)
+	c.Assert(asState, Equals, "")
+}
+
+func (s *deviceMgrSuite) TestHandleAutoImportAssertionFailed(c *C) {
+	a := devicestate.MockProcessAutoImportAssertion(func(*state.State, seed.Seed, asserts.RODatabase, func(batch *asserts.Batch) error) error {
+		return fmt.Errorf("failed to add user from system user assertions")
+	})
+	defer a()
+
+	s.mockSystemMode(c, "run")
+
+	s.state.Lock()
+	s.cacheDeviceCore20Seed(c)
+	s.seeding()
+	s.state.Unlock()
+
+	logbuf, restore := logger.MockLogger()
+	defer restore()
+
+	devicestate.EnsureAutoImportAssertions(s.mgr)
+
+	// ensure state has not been changed
+	s.state.Lock()
+	defer s.state.Unlock()
+	var asState string
+	err := s.state.Get("auto-as-imported", &asState)
+	c.Assert(err, IsNil)
+	c.Assert(asState, Equals, "done")
+	c.Assert(logbuf.String(), testutil.Contains, `failed to add user from system user assertions`)
+}
+
+func (s *deviceMgrSuite) TestHandleAutoImportAssertionMissingSerialAss(c *C) {
+	a := devicestate.MockProcessAutoImportAssertion(func(*state.State, seed.Seed, asserts.RODatabase, func(batch *asserts.Batch) error) error {
+		return devicestate.ErrDeviceNotRegistered
+	})
+	defer a()
+
+	s.mockSystemMode(c, "run")
+
+	s.state.Lock()
+	s.cacheDeviceCore20Seed(c)
+	s.seeding()
+	s.state.Unlock()
+
+	devicestate.EnsureAutoImportAssertions(s.mgr)
+
+	// check state is set as pending
+	s.state.Lock()
+	defer s.state.Unlock()
+	var asState string
+	err := s.state.Get("auto-as-imported", &asState)
+	c.Assert(err, IsNil)
+	c.Assert(asState, Equals, "pending")
+}
+
+func (s *deviceMgrSuite) TestHandleAutoImportAssertionWhileNotSeeding(c *C) {
+	a := devicestate.MockProcessAutoImportAssertion(func(*state.State, seed.Seed, asserts.RODatabase, func(batch *asserts.Batch) error) error {
+		return devicestate.ErrDeviceNotRegistered
+	})
+	defer a()
+
+	s.mockSystemMode(c, "run")
+
+	s.state.Lock()
+	s.cacheDeviceCore20Seed(c)
+	s.state.Unlock()
+
+	devicestate.EnsureAutoImportAssertions(s.mgr)
+
+	// ensure state has not been changed
+	s.state.Lock()
+	defer s.state.Unlock()
+	var asState string
+	err := s.state.Get("auto-as-imported", &asState)
+	c.Assert(err, testutil.ErrorIs, state.ErrNoState)
+	c.Assert(asState, Equals, "")
+}
+
+func (s *deviceMgrSuite) TestHandleAutoImportAssertionHappy(c *C) {
+
+	s.mockSystemMode(c, "run")
+
+	s.state.Lock()
+	s.cacheDeviceCore20Seed(c)
```

Comment:
at leas some tests need to deal with the interactions between ensureSeeded and ensureAutoImportAssertions because I think the code actually won't quite work as is afaict in all scenario, it's because of those interactions. One issue is the resetting of early seed, but there might be others.

Reply:
I do not think this is avoidable, ensure function needs lock in order to manipulate state key values. Tests at the same time need to validate right value is set for the key, so it needs lock.
Preparing at least minimal seed for testing also requires lock. I do not think there is easy way to avoid this lock management.

Reply:
@pedronis I do not think I follow here.
once accessing of the early seed is moved to be done under lock, I think this addresses this situation.
I have also updated test fort this.
I believe all the scenarios are tested there. Use of the caches seed is opportunistic and if early seed is not caches, we simply do not process it. In all the tests I have performed so far, this seems to work as expected, for both normal and preseeded scenarios. Auto import assertion is always processed correctly.

Test do need to deal with lock for simple reason, `ensureSeeded` requires lock in order to load and cache seed, I do not think this is avoidable, and it also matches real world scenario where seed is loaded under the lock.

Reply:
as I wrote I'm surprised that it works in the non-preseeded because of the resetting of the cache that I pointed but maybe there's an extra restart of snapd that happens, anyway it feels fragile to rely on behavior that is orthogonal to the code itself my preference is still to have test that go through Ensure and make sure the code works directly. For the latter we probably need to move the point where we reset earlyDeviceSeed.

Reply:
I did change some of the tests to go through Ensure